File: src/kreo_kontrol/device/domains/keymap.py
from __future__ import annotations

from typing import TypedDict

from pydantic import BaseModel


class KeyAction(BaseModel):
    action_id: str
    label: str
    category: str
    raw_value: int


class KeyActionOption(BaseModel):
    action_id: str
    label: str
    category: str
    raw_value: int


class KeyAssignment(BaseModel):
    ui_key: str
    logical_id: str
    svg_id: str
    label: str
    protocol_pos: int
    base_action: KeyAction
    fn_action: KeyAction


class KeymapState(BaseModel):
    verification_status: str
    assignments: list[KeyAssignment]
    available_actions: list[KeyActionOption]


class KeymapPayload(TypedDict):
    verification_status: str
    assignments: list[dict[str, object]]
    available_actions: list[dict[str, object]]

File: frontend/src/lib/keymap.ts
export interface KeyActionModel {
  actionId: string;
  label: string;
  category: string;
  rawValue: number;
}

export interface KeyAssignmentModel {
  uiKey: string;
  logicalId: string;
  svgId: string;
  label: string;
  protocolPos: number;
  baseAction: KeyActionModel;
  fnAction: KeyActionModel;
}

export interface KeymapModel {
  verificationStatus: string;
  assignments: KeyAssignmentModel[];
  availableActions: KeyActionModel[];
}

export interface KeymapEditPayload {
  base_raw_value?: number;
  fn_raw_value?: number;
}

const DEFAULT_KEY_COLOR = "#273240";
const REMAPPED_KEY_COLOR = "#2d6a5f";
const STAGED_KEY_COLOR = "#355c7d";

function isBaseAssignmentRemapped(assignment: KeyAssignmentModel): boolean {
  return assignment.baseAction.actionId !== `basic:${assignment.uiKey}`;
}

export function buildKeymapColorsBySvgId(
  assignments: KeyAssignmentModel[],
  stagedEdits: Record<string, KeymapEditPayload>,
): Map<string, string> {
  return new Map(
    assignments.map((assignment) => {
      if (stagedEdits[assignment.uiKey]) {
        return [assignment.svgId, STAGED_KEY_COLOR];
      }

      return [
        assignment.svgId,
        isBaseAssignmentRemapped(assignment) ? REMAPPED_KEY_COLOR : DEFAULT_KEY_COLOR,
      ];
    }),
  );
}
